### PR TITLE
Move to Flux Correction in One

### PR DESCRIPTION
**Pull request by lroberts36**
## PR Summary

Switch from per buffer loads and unloads of flux corrections to batched loads and unloads. This is in anticipation of generalizing the boundary values machinery to work with face, edge, and vertex centered fields. Also refactors bvals code since there was a lot of code duplication between the boundary communication routines and the flux correction in one routines.

## PR Checklist

- [x] Code passes cpplint
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/bvals/cc/boundary_communication.cpp
Comment (Yurlungur) [inline src/bvals/cc/boundary_communication.cpp:224]: Are tuple comprehensions a C++17 feature? Because that is really nice.

**Review by Yurlungur — APPROVED** [review]
This all seems like a straightforward re-org/re-name + extension of existing machinery. :+1:

File: src/bvals/cc/boundary_communication.cpp
Comment (lroberts36) [reply src/bvals/cc/boundary_communication.cpp:224]: Yeah, although they are not actually necessary, structured bindings are one of the C++17 features I was waiting on most.

File: src/bvals/cc/flux_correction.cpp
Comment (Yurlungur) [inline src/bvals/cc/flux_correction.cpp]: Should delete this.

File: src/bvals/cc/bnd_info.cpp
Comment (forrestglines) [inline src/bvals/cc/bnd_info.cpp:388]: There's a lot of repeated code for these types of sections that we've just sorted of accepted. I'm not suggesting to change these in this commit, but something to think about later.

File: src/bvals/cc/boundary_communication.cpp
Comment (forrestglines) [inline src/bvals/cc/boundary_communication.cpp:115]: We do this sort of indexing more than once but it's fairly error prone to duplicate. Could we make a template to do it for us? I suppose that would be difficult though due to the intermediate integers that we're trying to cache.

File: src/bvals/cc/bnd_info.cpp
Comment (forrestglines) [inline src/bvals/cc/bnd_info.cpp:254]: Was it a point of conversation during the Parthenon workshop to get rid of/avoid naked pointers from now on?

**Review by forrestglines — APPROVED** [review]
Looks good - some of this code might end up either duplicated or extended for Edge corrections. We'll see how that works out once more of the face centered machinery is in place.

File: src/bvals/cc/bnd_info.cpp
Comment (lroberts36) [reply src/bvals/cc/bnd_info.cpp:388]: Yeah, it would be good to think about if there is a way to unify these index calculation routines in a relatively simple way. They are all slightly different though, so it will require a little bit of thought.

File: src/bvals/cc/boundary_communication.cpp
Comment (lroberts36) [reply src/bvals/cc/boundary_communication.cpp:115]: I think the `parfor` abstractions do what you are suggesting. Here the loop structures are a little bit different, so I think it is easier to do by hand.

File: src/bvals/cc/bnd_info.cpp
Comment (forrestglines) [reply src/bvals/cc/bnd_info.cpp:388]: I think it would be straightforward to do iteratively if we had equivalent arrays for `out.si, out.sj, out.sk` and `inner_x1, inner_x2, inner_x3` etc. However, we've inherited a tendency from Athena++ to name out each different dimension. Still just a thought though.

File: src/bvals/cc/boundary_communication.cpp
Comment (forrestglines) [reply src/bvals/cc/boundary_communication.cpp:115]: The difference here is that we need both the flat index and the shaped indices to copy `var` into `buf`. I had in mind implementing something like [`numpy.unravel_index`](https://numpy.org/doc/stable/reference/generated/numpy.unravel_index.html). However, if we naively implemented that it we'd be recomputing the intermediate integers on each vector lane instead of once per thread.

File: src/bvals/cc/bnd_info.cpp
Comment (Yurlungur) [reply src/bvals/cc/bnd_info.cpp:254]: I don't believe so. I think naked pointers are still pretty common---I think the style is do naked pointers when passing into a function that does not modify pointer scope?

Comment (lroberts36) [reply src/bvals/cc/bnd_info.cpp:254]: yes, I think the standard wisdom is to pass raw pointers to a function if the function does not cause something to take ownership of the resource for longer than the lifetime of the function itself.

File: src/bvals/cc/flux_correction.cpp
Comment (lroberts36) [inline src/bvals/cc/flux_correction.cpp]:
```suggestion
```

Comment (lroberts36) [reply src/bvals/cc/flux_correction.cpp]: done

**Comment by lroberts36** [comment]
CI is failing with 
```
Error: EACCES: permission denied, open '/__w/_temp/_runner_file_commands/save_state_1b372095-2ad0-4072-9083-48fd89da0d52'
```
I am not sure what is going on here. Previously all tests were passing and the only change in the latest commit is to remove a single `printf` statement. @pgrete @yurlungur any ideas?

Also, I see that @pgrete turned off auto-merge. Did you want to do a review before merging?

**Comment by Yurlungur** [comment]
I have no idea... These are the github runners right? Maybe the host machine's filesystem is just choked and a rerun would resolve it?

**Comment by pgrete** [comment]
The failed pipelines/runners is on me. I started the transition to the new CI machine yesterday and am still fixing hiccups.

I disabled auto-merge last week as there were several unresolved discussion at that point were (as far as I read them) feedback was still requested.
Given that the discussion are now marked as resolved, I'll reenable auto-merge and retrigger the pipeline (once I got that fixed).

**Comment by pgrete** [comment]
Just saw that there's no changelog even though it's marked of in the checklist above.
Are all intended files in this PR or did sth go wrong with a rebase to `develop`?

**Comment by lroberts36** [comment]
Hm, I am not sure why the changelog doesn't show as being updated. I will take a look to see if something went wrong with the rebase.

**Comment by lroberts36** [comment]
I must have just erroneously clicked that box, everything else seems to be there. The changelog is updated now and the PR should be good to go once it passes the checks.

**Pull request merged**
